components/script/dom/htmlimageelement.rs

```diff
@@ -267,33 +267,23 @@ impl HTMLImageElement {
     fn fetch_image(&self, img_url: &ServoUrl) {
         let window = window_from_node(self);
         let image_cache = window.image_cache();
-        let response = image_cache.find_image_or_metadata(
-            img_url.clone().into(),
-            UsePlaceholder::Yes,
-            CanRequestImages::Yes,
-        );
-        match response {
-            Ok(ImageOrMetadataAvailable::ImageAvailable(image, url)) => {
-                self.process_image_response(ImageResponse::Loaded(image, url));
-            },
+        let cache_result =
+            image_cache.track_image(img_url.clone(), UsePlaceholder::Yes, CanRequestImages::Yes);
 
-            Ok(ImageOrMetadataAvailable::MetadataAvailable(m)) => {
-                self.process_image_response(ImageResponse::MetadataLoaded(m));
+        match cache_result {
+            ImageCacheResult::Available(ImageOrMetadataAvailable::ImageAvailable(image, url)) => {
+                self.process_image_response(ImageResponse::Loaded(image, url))
             },
-
-            Err(ImageState::Pending(id)) => {
-                add_cache_listener_for_element(image_cache, id, self);
+            ImageCacheResult::Available(ImageOrMetadataAvailable::MetadataAvailable(m)) => {
+                self.process_image_response(ImageResponse::MetadataLoaded(m))
             },
-
-            Err(ImageState::LoadError) => {
-                self.process_image_response(ImageResponse::None);
-            },
-
-            Err(ImageState::NotRequested(id)) => {
+            ImageCacheResult::Pending(id) => add_cache_listener_for_element(image_cache, id, self),
+            ImageCacheResult::ReadyForRequest(id) => {
                 add_cache_listener_for_element(image_cache, id, self);
```

Comment:
I think the idea I had is that the track_image API would automatically call ImageCache::add_listener so that users of the API would not have to do that.